src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,331 @@
+package consistenthashing
+
+import (
+	"context"
+	"hash/crc32"
+	"path"
+	"sort"
+	"sync"
+	"time"
+
+	"go.etcd.io/etcd/client/v3/concurrency"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/gogo/protobuf/types"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/collection"
+	"github.com/pachyderm/pachyderm/v2/src/internal/dlock"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+	"github.com/pachyderm/pachyderm/v2/src/internal/watch"
+
+	"github.com/sirupsen/logrus"
+
+	etcd "go.etcd.io/etcd/client/v3"
+)
+
+var (
+	// these variables are overridable for testing purposes.
+	hashFn = crc32.ChecksumIEEE
+	idGen  = uuid.New
+)
+
+// Ring is a consistent hash Ring. Each process should only create one instance with a given prefix. Nodes can be
+// added locally to a Ring. A Ring watches for changes to the prefix to determine if new nodes have been added
+// or deleted.
+type Ring struct {
+	client    *etcd.Client
+	hashFn    func([]byte) uint32
+	stateLock sync.Mutex
+	logger    *logrus.Logger
+	members   []member
+	node      node // node represents the member instance in a Ring that is local to the current process.
+	prefix    string
+}
+
+// node is a local member of a Ring. Upon creation, a node requests a lease from etcd and continuously refreshes it
+// until it is shutdown.
+type node struct {
+	member
+	locks map[string]lockInfo
+}
+
+// member contains metadata about members in a Ring.
+type member struct {
+	Id   string
+	hash uint32
+}
+
+// lockInfo instances are created for keys by a ring's node.
+// A context lockInfo.ctx is stored so the proper context is used when unlocking lockInfo.lock
+type lockInfo struct {
+	lock dlock.DLock
+	ctx  context.Context
+}
+
+func (ring *Ring) Members() []member {
+	return ring.members
+}
+
+func WithRing(ctx context.Context, client *etcd.Client, logger *logrus.Logger, prefix string, cb func(ctx context.Context, ring *Ring) error) error {
+	cancelCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	eg, ctx := errgroup.WithContext(cancelCtx)
+	id := idGen()
+	ring := ring(client, logger, prefix, id)
+	etcdCol := collection.NewEtcdCollection(ring.client, path.Join(ring.prefix, "nodes"), nil, nil, nil, nil)
+	eg.Go(func() error { return ring.watch(ctx, cancel, etcdCol) })
```

Comment:
The cancels should be removed from the two background goroutines. If they return an error, the errgroup context will be canceled. The cancel context we create is only for indicating successful completion of the client provided callback.